UnityProject/Assets/Scripts/UI/GUI_Vendor.cs

```diff
@@ -1,179 +1,113 @@
-	[SerializeField]
-	private float cooldownTimer = 2f;
-
-	private Vendor vendor;
-	private List<VendorItem> vendorContent = new List<VendorItem>();
 	[SerializeField]
 	private EmptyItemList itemList = null;
 	[SerializeField]
 	private NetColorChanger hullColor = null;
-	private bool inited = false;
-	[SerializeField]
-	private string deniedMessage = "Bzzt.";
-	[SerializeField]
-	private string restockMessage = "Items restocked.";
 
-	private void Start()
-	{
-		if (!CustomNetworkManager.Instance._isServer)
-		{
-			return;
-		}
-	}
+	private Vendor vendor;
 
 	protected override void InitServer()
 	{
 		StartCoroutine(WaitForProvider());
 	}
 
-	IEnumerator WaitForProvider()
+	private IEnumerator WaitForProvider()
 	{
 		while (Provider == null)
 		{
+			// waiting for Provider
 			yield return WaitFor.EndOfFrame;
 		}
-		vendor = Provider.GetComponent<Vendor>();
-		hullColor.SetValueServer(vendor.HullColor);
-		inited = true;
-		GenerateContentList();
-		UpdateList();
-		vendor.OnRestockUsed.AddListener(RestockItems);
-	}
 
-	private void GenerateContentList()
-	{
-		if (!CustomNetworkManager.Instance._isServer)
+		vendor = Provider.GetComponent<Vendor>();
+		if (vendor)
 		{
-			return;
-		}
+			hullColor.SetValueServer(vendor.HullColor);
+			UpdateAllItemsView();
 
-		vendorContent = new List<VendorItem>();
-		for (int i = 0; i < vendor.VendorContent.Count; i++)
-		{
-			//protects against missing references
-			if (vendor.VendorContent[i] != null && vendor.VendorContent[i].Item != null)
-			{
-				vendorContent.Add(new VendorItem(vendor.VendorContent[i]));
-			}
+			vendor.OnItemVended.AddListener(UpdateItemView);
+			vendor.OnRestockUsed.AddListener(UpdateAllItemsView);
 		}
 	}
 
 	public override void OnEnable()
 	{
 		base.OnEnable();
-		if (!CustomNetworkManager.Instance._isServer || !inited)
+		if (CustomNetworkManager.IsServer)
 		{
-			return;
+			UpdateAllItemsView();
 		}
-		UpdateList();
-		allowSell = true;
 	}
 
-	public void RestockItems()
+	/// <summary>
+	/// Buy UI button was pressed by client
+	/// </summary>
+	public void OnVendItemButtonPressed(VendorItem vendorItem, ConnectedPlayer player)
 	{
-		GenerateContentList();
-		UpdateList();
-		SendToChat(restockMessage);
+		if (vendor)
+		{
+			vendor.TryVendItem(vendorItem, player);
+		}
 	}
 
-	private void UpdateList()
+	/// <summary>
+	/// Clear all items and send new vendor state to clients UI
+	/// </summary>
+	private void UpdateAllItemsView()
 	{
+		if (!vendor)
+		{
+			return;
+		}
+
+		// remove all items UI
 		itemList.Clear();
+
+		var vendorContent = vendor.VendorContent;
 		itemList.AddItems(vendorContent.Count);
+
+		// update UI for clients
 		for (int i = 0; i < vendorContent.Count; i++)
 		{
 			VendorItemEntry item = itemList.Entries[i] as VendorItemEntry;
 			item.SetItem(vendorContent[i], this);
 		}
 	}
 
-	public void VendItem(VendorItem item)
+	/// <summary>
+	/// Update only single item and send new state to clients UI
```

Comment:
This one fixes annoying bug when you selected some item to buy and lists sliders starts moving like crazy. It's because UI dynamically download and rebuild same entities again client side (can be hundred in one vendor). This function doesn't remove any elements, but only update single one values.

Not sure how to encapsulate this function to dynamic list. It updates values using SetItem function which will be different for each list application. Though I'm not very familiar with NetUI yet, so it may be made generic somehow.